src/PowerShellAPI.cs

```diff
@@ -0,0 +1,143 @@
+                    throw new CommandNotFoundException(commandName);
+                }
+            }
+            // This happens when Get-Command throws and ErrorAction is set to Stop
+            catch (System.Management.Automation.ActionPreferenceStopException apse)
+            {
+                if (apse.ErrorRecord.Exception is CommandNotFoundException cnfe)
+                {
+                    throw cnfe;
+                }
+
+                throw;
+            }
+        }
+
+        public static Collection<CommandInfo> GetCmdletInfoFromModule(string moduleName)
+        {
+            ps ??= System.Management.Automation.PowerShell.Create(RunspaceMode.CurrentRunspace);
+            ps.Commands.Clear();
+
+            Collection<PSModuleInfo>? moduleInfo = GetModuleInfo(moduleName);
+
+            Collection<CommandInfo> cmdletInfos = new();
+
+            if (moduleInfo?.Count > 0)
+            {
+                foreach (var mod in moduleInfo)
+                {
+                    foreach (var cmdletInfo in mod.ExportedCommands.Values)
+                    {
+                        if (cmdletInfo.CommandType != CommandTypes.Alias)
+                        {
+                            cmdletInfos.Add(cmdletInfo);
+                        }
+                    }
+                }
+            }
+            else
+            {
+                throw new ItemNotFoundException(moduleName);
+            }
+
+            return cmdletInfos;
+        }
+
+        public static Collection<PSModuleInfo>? GetModuleInfo(string moduleName)
+        {
+            if (string.IsNullOrEmpty(moduleName))
+            {
+                return null;
+            }
+
+            ps ??= System.Management.Automation.PowerShell.Create(RunspaceMode.CurrentRunspace);
+            ps.Commands.Clear();
+
+            Collection<PSModuleInfo>? modules = null;
+
+            try
+            {
+                // first try if the module is loaded
+                modules = ps
+                    .AddCommand(@"Microsoft.PowerShell.Core\Get-Module")
+                    .AddParameter("Name", moduleName)
+                    .Invoke<PSModuleInfo>();
+
+                if (modules.Count == 0)
+                {
+                    // if not found try to import first
+
+                    modules = ps
+                        .AddCommand(@"Microsoft.PowerShell.Core\Import-Module")
+                        .AddParameter("Name", moduleName)
+                        .AddParameter("PassThru")
+                        .Invoke<PSModuleInfo>();
+                }
+            }
+            catch (FileNotFoundException)
+            {
+                // swallow the exception and eventually return null;
+            }
+
+            return modules;
+        }
+
+        internal static void InitializeRemoteSession(PSSession session)
+        {
+            ps = System.Management.Automation.PowerShell.Create();
+            ps.Runspace = session.Runspace;
```

Comment:
Don't assign to the thread static instance. When a remote session is provided, we should create a new PowerShell instance just for the remote session. Otherwise, it will fail for the next local operation.

Reply:
Change the structure a bit. Now all the methods that the session and that decides whether we create a new one or use the thread local ps.